### PR TITLE
nonPayableFuncCheck hook

### PR DESCRIPTION
**Pull request by andrei-marinica**
New hook that checks that no payment was sent.
Checks both EGLD and ESDT.
Updated mandos tests to use it.

### CODE REVIEW

File: arwen/elrondapi/elrondei.go [hunk 1]
```diff
@@ -22,6 +22,7 @@ package elrondapi
 // extern int32_t storageLoad(void *context, int32_t keyOffset, int32_t keyLength , int32_t dataOffset);
 // extern int32_t storageLoadFromAddress(void *context, int32_t addressOffset, int32_t keyOffset, int32_t keyLength , int32_t dataOffset);
 // extern void getCaller(void *context, int32_t resultOffset);
+// extern void nonPayableFuncCheck(void *context);
```
Comment (andrei-marinica) [inline arwen/elrondapi/elrondei.go:25]: If you have a better name in mind, it's not too late to change

Comment (BeniaminDrasovean) [reply arwen/elrondapi/elrondei.go:25]: Maybe checkNonPayableFunc?

**Review by BeniaminDrasovean — APPROVED** [review]

File: arwen/elrondapi/elrondei.go (hunk 1 above)
Comment (andrei-marinica) [reply arwen/elrondapi/elrondei.go:25]: Yeah ... thought about that, but I don't want it to seem like we are checking whether or not the function is non payable. We are checking that no payment came.

Comment (sasurobert) [reply arwen/elrondapi/elrondei.go:25]: I would say checkFunctionNonPayable

File: arwen/elrondapi/elrondei.go [hunk 2]
```diff
@@ -1002,6 +1008,27 @@ func getCaller(context unsafe.Pointer, resultOffset int32) {
...
+//export nonPayableFuncCheck
+func nonPayableFuncCheck(context unsafe.Pointer) {
+	runtime := arwen.GetRuntimeContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	gasToUse := metering.GasSchedule().ElrondAPICost.GetCallValue
+	metering.UseGas(gasToUse)
+
+	vmInput := runtime.GetVMInput()
+	if vmInput.CallValue.Sign() > 0 {
```
Comment (sasurobert) [inline arwen/elrondapi/elrondei.go:1020]: vmInput.CallValue.Cmp(zero) > 0 as with ESDTValue -> but for esdt check for nil first.

File: arwen/elrondapi/elrondei.go (hunk 1 above)
Comment (BeniaminDrasovean) [reply arwen/elrondapi/elrondei.go:25]: Right, makes sense!

File: arwen/elrondapi/elrondei.go [hunk 3]
```diff
@@ -1002,6 +1008,27 @@ func getCaller(context unsafe.Pointer, resultOffset int32) {
...
+		runtime := arwen.GetRuntimeContext(context)
+		arwen.WithFault(arwen.ErrNonPayableFunctionEgld, context, runtime.ElrondAPIErrorShouldFailExecution())
+		return
+	}
+	if vmInput.ESDTValue.Sign() > 0 {
```
Comment (sasurobert) [inline arwen/elrondapi/elrondei.go:1025]: if vmInput.EsdtValue != nil && vmInput.ESDTValue.Cmp(zero) > 0

**Review by sasurobert — CHANGES_REQUESTED** [review]

**Review by sasurobert — APPROVED** [review]

**Review by BeniaminDrasovean — APPROVED** [review]

**Pull request merged**